nav2_util/include/nav2_util/simple_action_server.hpp

```diff
@@ -137,22 +176,45 @@ class SimpleActionServer
   {
     std::lock_guard<std::recursive_mutex> lock(update_mutex_);
     server_active_ = true;
+    stop_execution_ = false;
   }
 
   void deactivate()
   {
-    std::lock_guard<std::recursive_mutex> lock(update_mutex_);
-    server_active_ = false;
+    debug_msg("Deactivating...");
 
-    if (is_active(current_handle_)) {
-      warn_msg("Taking action server to deactive state with an active goal.");
+    {
+      std::lock_guard<std::recursive_mutex> lock(update_mutex_);
+      server_active_ = false;
+      stop_execution_ = true;
+    }
+
+    if (is_running()) {
+      warn_msg("Requested to deactivate server but goal is still executing."
+        " Should check if action server is running before deactivating.");
+    }
+
+
+    using namespace std::chrono;  //NOLINT
+
+    auto end_time = steady_clock::now();
```

Comment:
This is a little confusing. I'd change to:

`while (execution_future_.wait_for(milliseconds(100)) != std::future_status::ready && now() - start_time < server_timeout_ )`

then remove the `end_time` for `start_time = now()`

Reply:
I've simplified it based on your suggestion.

I was thinking more about what should happen if the thread missed the time to stop during deactivation... Maybe the managed node calling the deactivation should stop transitioning and go to an error state? I assume that will happen by throwing this exception.